OSDOCS-3887: add ELB switching feature (#50439)
openshift/openshift-docs: modules/nw-aws-switching-clb-with-nlb.adoc

// Module included in the following assemblies:
//
// * networking/configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-aws.adoc

:_content-type: PROCEDURE
[id="nw-aws-switching-clb-with-nlb_{context}"]
= Switching the Ingress Controller from using a Classic Load Balancer to a Network Load Balancer

You can switch the Ingress Controller that is using a Classic Load Balancer (CLB) to one that uses a Network Load Balancer (NLB) on AWS.

Switching between these load balancers will not delete the IngressController object and thus does not cause an Ingress interruption.

.Procedure

. Modify the existing Ingress Controller that you would like to switch to using a NLB. The following example assumes that your default Ingress Controller has an `External` scope and no other customizations:
+

> .Example `ingresscontroller.yml` file

Reviewer: Most documentation uses `.yaml`, but `.yml` is used here and some other places.  Either should work, and the `oc apply` command doesn't care how the file is named (as long as the actual file name is provided to `oc apply`), but the inconsistency caught my eye.

Author: Good catch, for consistency I'll update this to use the `.yaml` extension

[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  creationTimestamp: null
  name: default
  namespace: openshift-ingress-operator
spec:
  endpointPublishingStrategy:
    loadBalancer:
      scope: External
      providerParameters:
        type: AWS
        aws:
          type: NLB
    type: LoadBalancerService
----
+
If your default Ingress Controller has other customizations, ensure that you modify the file accordingly.

. Apply the changes to the Ingress Controller YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f ingresscontroller.yml
----
+
[NOTE]
====

> When you modify the Ingress Controller to use a NLB, the Ingress Controller retains the `platform.aws.lbType` value that was set during installation.

Reviewer: ```suggestion
If you do not specify `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.type`, the Ingress Controller uses the `spec.loadBalancer.platform.aws.type` value from the cluster `Ingress` configuration that was set during installation.
```

// Module included in the following assemblies:
//
// * networking/configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-aws.adoc

:_content-type: PROCEDURE
[id="nw-aws-switching-clb-with-nlb_{context}"]
= Switching the Ingress Controller from using a Classic Load Balancer to a Network Load Balancer

You can switch the Ingress Controller that is using a Classic Load Balancer (CLB) to one that uses a Network Load Balancer (NLB) on AWS.

> Switching between these load balancers will not delete the IngressController object.

Reviewer 2: should "the `IngressController` object" be backticks?

Author: Good catch! Yes it should be 🙂

[WARNING]
====

> This procedure might cause the following issues:

Reviewer 2: See comment above about `following`

* An outage that can last several minutes due to new DNS records propagation, new load balancers provisioning, and other factors. IP addresses and canonical names of the Ingress Controller load balancer might change after applying this procedure.

* Leaked load balancer resources due to a change in the annotation of the service.
====

.Procedure

> . Modify the existing Ingress Controller that you would like to switch to using an NLB. The following example assumes that your default Ingress Controller has an `External` scope and no other customizations:

Reviewer 2: ```suggestion
. Modify the existing Ingress Controller that you want to switch to using an NLB. The example assumes that your default Ingress Controller has an `External` scope and no other customizations:
```
See `want` in SSG for the replacement https://redhat-documentation.github.io/supplementary-style-guide/#_w also removed `following` based on my previous comment.

+
.Example `ingresscontroller.yaml` file
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  creationTimestamp: null
  name: default
  namespace: openshift-ingress-operator
spec:
  endpointPublishingStrategy:
    loadBalancer:
      scope: External
      providerParameters:
        type: AWS
        aws:
          type: NLB
    type: LoadBalancerService
----
+
[NOTE]
====
If you do not specify a value for the `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.type` field, the Ingress Controller uses the `spec.loadBalancer.platform.aws.type` value from the cluster `Ingress` configuration that was set during installation.
====
+
[TIP]
====
If your Ingress Controller has other customizations that you want to update, such as changing the domain, consider force replacing the Ingress Controller definition file instead.
====

> . Apply the changes to the Ingress Controller YAML file by running the following command:

Reviewer 2: I think you could remove "the following command". I will send you the slack thread I am thinking of. 
```suggestion
. Apply the changes to the Ingress Controller YAML file by running:
```

Reviewer 2: or 
```suggestion
. Apply the changes to the Ingress Controller YAML file by running the command:
```

+
[source,terminal]
----
$ oc apply -f ingresscontroller.yaml
----
+

> Wait until the Ingress Controller is updated. Expect several minutes of outages.

Reviewer 2: Just a suggestion, but what do you think of
```suggestion
Expect several minutes of outages while the Ingress Controller updates.
```

Author: I like it! Thank you
